enhancements/rhcos/complex-devices.md

```diff
@@ -0,0 +1,286 @@
+   that were not encrypted)
+ - Nested configurations
+
+## Proposal
+
+There are multiple different areas that will require changes to support this
+proposal.
+ - FCCT gains sugar that writes a file to `/boot` specifying how to mount root
+   devices
+ - Ignition gains support for LUKS devices (RAID already supported)
+ - ostree work to allow re-laying the ostree on new root devices from inside the
+   initrd
+ - New initrd module that runs very early in the boot process which can process
+   a file in `/boot` and determine how to mount the root device
+    * Could be as simple as passing through to Clevis for LUKS devices
+    * Or, use kernel arguments
+ - New initrd module that runs in between the Ignition Disks & Files stages that
+   lays down the ostree on new root devices (`ostree-redeploy-rootfs.service`)
+
+### Implementation Details/Notes/Constraints
+
+This implementation relies on a file written by Ignition in /boot to determine
+how to mount the root device on subsequent boots. If this file is incorrect the
+machine will fail to boot on subsequent boots.
+
+The easy way for users to create an Ignition configuration that specifies a new
+root device would be the use of FCCT which is outside of the OpenShift
+installer. While it's still possible to specify the file by hand it could be
+error-prone, leading to frustration as systems fail to mount the root device on
+subsequent boots.
+
+----
+
+Boot Order (first boot):
+ 1. Ignition disks
+     - Handles formatting disks / creating partitions, opening encrypted volumes (to allow
+       for the creation of filesystems on encrypted volumes), closing encrypted volumes
+ 2. Dracut module runs parsing /boot file & unlocks sysroot
+     - Needs to parse Ignition file for /boot file contents
+ 3. mount sysroot
+ 4. ostree-redeploy-rootfs
+ 5. Ignition mount
+     - Opens & mounts non-root volumes
+ 6. Ignition files
+ 7. Ignition umount
+     - Closes & umounts non-root volumes
+
+----
+
+Boot Order (subsequent boots):
+ 1. Dracut module runs parsing /boot file & unlocks sysroot
+ 2. mount sysroot
+
+----
+
+Other than the root filesystem everything else is opened / mounted by the booted
+system (e.x. `/etc/crypttab`). The distribution owns mounting & unlocking the
+root filesystem.
+
+----
+
+LUKS RHCOS currently has `rhcos.rootfs=luks`, use this to detect the upgrade case.
+
+#### Examples
+
+All examples are in FCC format (https://github.com/coreos/fcct).
+
+How to specify a new root device:
+```
+variant: rhcos
+version: 1.0.0
+storage:
+  disks:
+    - device: /dev/sda
+      wipe_table: true
+    - device: /dev/sdb
+      wipe_table: true
+      partitions:
+        - number: 1
+          label: root
+  filesystems:
+    - path: /
+      device: /dev/disk/by-partlabel/root
+      format: xfs
+      wipe_filesystem: true
+      contains_root: true
+      label: root
+```
+
+Creating a LUKS (KeyFile) device:
+```
+variant: rhcos
+version: 1.0.0
```

Comment:
Nit: if we want to keep the RHCOS and FCOS variants in sync, LUKS support will probably be in 1.2.0.